wgpu-hal/src/dx12/mod.rs

```diff
@@ -654,13 +654,18 @@ impl crate::Surface<Api> for Surface {
 
         let non_srgb_format = auxil::dxgi::conv::map_texture_format_nosrgb(config.format);
 
+        // Nvidia recommends to use 1-2 more buffers than the maximum latency
+        // https://developer.nvidia.com/blog/advanced-api-performance-swap-chains/
+        // For high latency extra buffers seems excessive, so go with a minimum of 3 and beyond that add 1.
+        let swap_chain_buffer = (config.maximum_frame_latency + 1).min(3);
```

Comment:
won't this make double-buffering impossible?

Reply:
The number of swap chain buffer isn't _actually_ the thing we care about to begin with: we care about the latency from recording to display and DX12 can set this more directly than using [`SetMaximumFrameLatency`](https://learn.microsoft.com/en-us/windows/win32/api/dxgi/nf-dxgi-idxgidevice1-setmaximumframelatency). Essentially it decouples the "wait on queue is full" from the number of buffers that are used. Which is why @cwfitzgerald actually advocated originally to just always use 3. But naturally, the desired frequency isn't entirely decoupled which is why Nvidia suggests to over provision by at least 1.
To my understanding the major perk of this more principled approach is that instead of blocking on `Present`/buffer flip, y we wait at the begin of a frame using a "waitable object" which is signaled **after** a present is done. See [getframelatencywaitableobject](https://learn.microsoft.com/en-us/windows/win32/api/dxgi1_3/nf-dxgi1_3-idxgiswapchain2-getframelatencywaitableobject). This means that if there's (for whatever reason) a longer wait, you have frame preparation including all the input query much much closer to the actual presenting of the frame.

Timing your frame preparation more accurately and consistently gets you into the even more complex real of "frame pacing" which is all about pacing your frame recording to hit the next screen refresh as closely as possible; there's various mutally inconsistent APIs popping up these days (see #2869) so it's that part of the equation isn't that easy to solve.

Unfortunately, on Vulkan this works only with some extensions. To be implemented. (Oh and since you're wondering now: Rerun doesn't use DX12 right now for a bunch or reasons)

I'm still learning this stuff, so maybe @cwfitzgerald can correct me if I got something wrong :)

Reply:
tl'dr: The trick is that on dx12 `SetMaximumFrameLatency` takes `maximum_frame_latency` as-is.

Reply:
…I also misread `.min(3)` as "at minimum 3" 🤦